File: UPCODER/Other/BSTree_TEST.cpp
#include <iostream>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int value) : data(value), left(nullptr), right(nullptr) {}
};

Node* insert(Node* root, int data) {
    if (root == nullptr) 
        return new Node(data);
        
    if (data < root->data) 
        root->left = insert(root->left, data);
    else if (data > root->data) 
        root->right = insert(root->right, data);

    return root;
}

void inOrder(Node* root) {
    if (root != nullptr) {
        inOrder(root->left);
        cout << root->data << " ";
        inOrder(root->right);
    }
}

void preOrder(Node* root) {
    if (root != nullptr) {
        cout << root->data << " ";
        preOrder(root->left);
        preOrder(root->right);
    }
}

void postOrder(Node* root) {
    if (root != nullptr) {
        postOrder(root->left);
        postOrder(root->right);
        cout << root->data << " ";
    }
}

int main() {
    int tmp, n;
    Node* root = nullptr;
    
    cin >> n;
    while (n--) {
        cin >> tmp; 
        root = insert(root, tmp);
    }

    // Left-Node-Right (Duyệt theo thứ tự)
    inOrder(root);
    cout << endl;

    // Node-Left-Right (Duyệt trước)
    preOrder(root);
    cout << endl;

    // Left-Right-Node (Duyệt sau)
    postOrder(root);
    return 0;
}

File: UPCODER/Other/BSTree_TEST.py
class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

def insert(root, data):
    if root is None:
        return Node(data)
    if data < root.data:
        root.left = insert(root.left, data)
    elif data > root.data:
        root.right = insert(root.right, data)
    return root

def inOrder(root):
    if root is not None:
        inOrder(root.left)
        print(root.data, end=' ')
        inOrder(root.right)

def preOrder(root):
    if root is not None:
        print(root.data, end=' ')
        preOrder(root.left)
        preOrder(root.right)

def postOrder(root):
    if root is not None:
        postOrder(root.left)
        postOrder(root.right)
        print(root.data, end=' ')

n = int(input())
list = list(map(int, input().split()))
root = None
for tmp in list:
    root = insert(root, tmp)

inOrder(root)
print()

preOrder(root)
print()

postOrder(root)